packages/components/src/shortcuts/Shortcut.ts

```diff
@@ -0,0 +1,324 @@
+   * @returns Type predicate asserting the key is an allowed KEY
+   */
+  static isAllowedKey(key: string | null): key is KEY {
+    return Object.values(KEY).includes(key as KEY);
+  }
+
+  /**
+   * Checks if a KeyState has a valid key.
+   * @param state KeyState with any string as the key
+   * @returns True if KeyState is is using an allowed keyCode
+   */
+  static isValidKeyState(state: KeyState): state is ValidKeyState {
+    return (
+      Shortcut.isAllowedKey(state.keyValue) &&
+      (state.altKey ||
+        state.ctrlKey ||
+        state.metaKey ||
+        state.shiftKey ||
+        state.keyValue === KEY.ENTER ||
+        state.keyValue === KEY.DELETE)
+    );
+  }
+
+  static isMacPlatform = ContextActionUtils.isMacPlatform();
+
+  /**
+   * Creates a KeyState from a valid array of modifier and key constants
+   * @param keys Array of keys in the shortcut. Modifiers first and key last. Should use the MODIFIER and KEY enums
+   * @returns KeyState representing the array of constants
+   */
+  static createKeyState(keys: ShortcutKeys): ValidKeyState {
+    const keyState: ValidKeyState = {
+      altKey: false,
+      ctrlKey: false,
+      metaKey: false,
+      shiftKey: false,
+      keyValue: keys[keys.length - 1] as KEY,
+    };
+
+    keys.forEach(key => {
+      switch (key) {
+        case MODIFIER.CTRL:
+          keyState.ctrlKey = true;
+          break;
+        case MODIFIER.ALT:
+        case MODIFIER.OPTION:
+          keyState.altKey = true;
+          break;
+        case MODIFIER.CMD:
+          keyState.metaKey = true;
+          break;
+        case MODIFIER.SHIFT:
+          keyState.shiftKey = true;
+          break;
+        default:
+          break;
+      }
+    });
+
+    return keyState;
+  }
+
+  static getKeyStateFromEvent(
+    e: React.KeyboardEvent | KeyboardEvent
+  ): KeyState {
+    const { key: eventKey, keyCode } = e;
+    let key = '';
+    if (Shortcut.isAllowedKey(eventKey)) {
+      key = eventKey;
+    } else if (Shortcut.isAllowedKey(String.fromCharCode(keyCode))) {
+      key = String.fromCharCode(keyCode);
+    }
+
+    return {
+      keyValue: key,
+      altKey: e.altKey,
+      ctrlKey: e.ctrlKey,
+      metaKey: e.metaKey,
+      shiftKey: e.shiftKey,
+    };
+  }
+
+  /**
+   * Gets display string for Windows (and Linux + any other non-Mac OS)
+   * If the key is not an allowed key, the display will only display the modifiers
+   * @param keyState KeyState to get the display for
+   * @returns The string to display on Windows/non-Mac OS
+   */
+  private static getWindowsDisplayText(keyState: KeyState): string {
+    let display = '';
+
+    if (keyState.ctrlKey) {
+      display += 'Ctrl+';
+    }
+    if (keyState.altKey) {
+      display += 'Alt+';
+    }
+    if (keyState.shiftKey) {
+      display += 'Shift+';
+    }
+
+    if (keyState.keyValue === KEY.ESCAPE) {
+      display += 'Esc';
+    } else if (Shortcut.isAllowedKey(keyState.keyValue)) {
```

Comment:
If the key isn't allowed - should we throw in that case? Or fallback to just appending the keyvalue anyway for the display text? Not sure why this check is here

Reply:
Appends the keyvalue now so it displays the full invalid shortcut